presto-main/src/main/java/com/facebook/presto/operator/GroupedTopNBuilder.java

```diff
@@ -0,0 +1,654 @@
+        }
+
+        public void reset(int position)
+        {
+            this.position = position;
+        }
+
+        public int getPageId()
+        {
+            return pageId;
+        }
+
+        public int getPosition()
+        {
+            return position;
+        }
+    }
+
+    private static class PageReference
+    {
+        private static final long INSTANCE_SIZE = ClassLayout.parseClass(PageReference.class).instanceSize();
+
+        private Page page;
+        private Row[] reference;
+
+        private int usedPositionCount;
+
+        public PageReference(Page page)
+        {
+            reset(requireNonNull(page, "page is null"), new Row[page.getPositionCount()]);
+        }
+
+        public void reference(Row row)
+        {
+            int position = row.getPosition();
+            reference[position] = row;
+            usedPositionCount++;
+        }
+
+        public void dereference(int position)
+        {
+            checkArgument(reference[position] != null && usedPositionCount > 0);
+            reference[position] = null;
+            usedPositionCount--;
+        }
+
+        public int getUsedPositionCount()
+        {
+            return usedPositionCount;
+        }
+
+        public void compact()
+        {
+            checkState(usedPositionCount > 0);
+
+            if (usedPositionCount == page.getPositionCount()) {
+                return;
+            }
+
+            // re-assign reference
+            Row[] newReference = new Row[usedPositionCount];
+            int[] positions = new int[usedPositionCount];
+            int index = 0;
+            for (int i = 0; i < page.getPositionCount(); i++) {
+                if (reference[i] != null) {
+                    newReference[index] = reference[i];
+                    positions[index] = i;
+                    index++;
+                }
+            }
+            verify(index == usedPositionCount);
+
+            // compact page
+            Block[] blocks = new Block[page.getChannelCount()];
+            for (int i = 0; i < page.getChannelCount(); i++) {
+                Block block = page.getBlock(i);
+                blocks[i] = block.copyPositions(positions, 0, usedPositionCount);
+            }
+
+            // update all the elements in the heaps that reference the current page
+            for (int i = 0; i < usedPositionCount; i++) {
+                // this does not change the elements in the heap;
+                // it only updates the value of the elements; while keeping the same order
+                newReference[i].reset(i);
+            }
+            reset(new Page(usedPositionCount, blocks), newReference);
+        }
+
+        public Page getPage()
+        {
+            return page;
+        }
+
+        public long getEstimatedSizeInBytes()
+        {
+            return page.getRetainedSizeInBytes() + sizeOf(reference) + INSTANCE_SIZE;
+        }
+
+        private void reset(Page page, Row[] reference)
```

Comment:
I'd inline this into the two places... it made me think that the page could be reset externally (I see it is private but inner classes break access rules).